### PR TITLE
Start adding spec-compliant types

### PR DESCRIPTION
**Pull request by gjtrowbridge**
Originally, this simulation's MVP was built based on very simple, non-spec-compliant representations of eth2 state.  This diff begins the process of migrating to spec-compliant representations of state.

All added logic is based heavily on the spec-compliant (and very thorough) stuff found in the [sigp/lighthouse repo](https://github.com/sigp/lighthouse).  This simulation does not intend to be a fully spec-compliant Eth2 node (which I believe is lighthouse's goal), but it does intend to be spec-compliant for all eth2 structs, and to (long term) approximate the behavior of an Eth2 beacon chain + shard chains.

This diff adds:
* `BeaconState`
* `ExecutionEnvironment`
* `CrossLink`
* supporting logic to make the above structs spec-compliant
* supporting logic to make the above structs SSZ-serializable

Future diffs will address the following (not necessarily in this order):
* Add additional spec-compliant types as necessary
* Add the external-facing methods defined in the [roadmap](https://github.com/quilt/simulation/issues/21) where they don't already exist
* Hook up the new spec-compliant types to the internal state of the simulation, remove the "old" non-spec-compliant representations.
  * (internal state of the simulation may not exactly mirror the spec, but all specced representations (eg. `BeaconState`, `BeaconBlock`, etc) should be spec-compliant)
* Add merkleization support for Eth2 types
* Eventually add more true-to-life simulation under the hood (eg. creating blocks and transitioning state of `BeaconState` as a result of incoming blocks vs. just having `BeaconState`)

### CODE REVIEW

File: eth2/types/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,24 @@
+[package]
+name = "types"
```
Comment (SamWilsn) [inline eth2/types/Cargo.toml:2]: `types` is already an [existing crate](https://crates.io/crates/types). If we ever want to publish to crates.io, this might be an issue.

File: eth2/utils/int_to_bytes/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,10 @@
+// Copied from sigp/lighthouse with some modifications and deletions
```
Comment (SamWilsn) [inline eth2/utils/int_to_bytes/src/lib.rs]: sigp/lighthouse is licensed under Apache-2.0, which means you need to include the license document, at the very least. More information: https://tldrlegal.com/license/apache-license-2.0-(apache-2.0)

**Review by SamWilsn — COMMENTED** [review]
So I'm not sure how I feel about copying the files from lighthouse directly. Seems odd that they didn't pick unique crate names and just publish them to crates.io...

Can we use git crates instead? Something like:

```
[dependencies.foobar]
git = "https://github.com/sigp/lighthouse"
```

If that doesn't work, you're good to merge as far as I'm concerned.

**Comment by gjtrowbridge** [comment]
> So I'm not sure how I feel about copying the files from lighthouse directly. Seems odd that they didn't pick unique crate names and just publish them to crates.io...
> 
> Can we use git crates instead? Something like:
> 
> ```
> [dependencies.foobar]
> git = "https://github.com/sigp/lighthouse"
> ```
> 
> If that doesn't work, you're good to merge as far as I'm concerned.

Definitely a fair concern.  Here's my take:
1) In most cases, the "copy" from lighthouse actually required making a number of changes to make things work for the simulation.  In many cases I also heavily deleted (or commented) logic that wasn't applicable to the goals of the simulation (at least at the moment).
2) There are a number of crates already pushed by the Lighthouse folks.  Where possible, I used those crates instead of copying in code or writing my own code.  In some cases, those crates didn't completely match the code in lighthouse (eg. slightly different trait on cargo than was used in the `master` repo of lighthouse), which is one of the reasons I had to modify some of the code I copied in from the lighthouse repo.
3) For some of the code (or at least patterns) that I ported over, I think it will be nice to be able to edit it as necessary going forward.
4) For other parts of the code, I think if more parts of lighthouse get pushed as crates in the future, we can switch over to using the crates instead, and as more code gets added to this repo I think we'll have a better sense of which things we want this repo to own vs. what things we can pull in as crates.

TL;DR: For now, I think it will be simpler and easier to be OK with copying some small amounts of logic (with credit given of course).  If/when more crates are pushed, we can move the logic that we don't want to own over to using those newly-pushed crates.  I also think as we write more code in the simulation, we'll get a better sense of which parts we'd like to have as crates vs. what stuff we want to edit/own in this repo.  I also don't want us to get blocked at this early stage by trying too hard to make everything come from a crate.

File: eth2/utils/int_to_bytes/src/lib.rs (hunk 2 above)
Comment (gjtrowbridge) [reply eth2/utils/int_to_bytes/src/lib.rs]: Awesome, thanks, I'll update this.

File: eth2/types/Cargo.toml (hunk 1 above)
Comment (gjtrowbridge) [reply eth2/types/Cargo.toml:2]: Fair enough, thanks for the heads up!  I think we can cross this bridge if/when we come to it -- seems fairly easy to change this later if we want to publish this as a crate (though I doubt we'd want to do that any time soon if ever, since the types in the simulation may in some ways be specific to the needs of the simulation, but I could be wrong).

File: eth2/types/src/eth_spec.rs [hunk 3]
```diff
@@ -0,0 +1,91 @@
+// Copied from file of the same name at https://github.com/sigp/lighthouse
+// with some modifications and deletions. Specifically, removed code that wasn't necessary for this
```
Comment (villanuevawill) [inline eth2/types/src/eth_spec.rs:2]: @gjtrowbridge do we need all these configurations? It seems like most of this would be irrelevant?

File: eth2/types/src/execution_environment.rs [hunk 4]
```diff
@@ -0,0 +1,11 @@
+use serde::{Deserialize, Serialize};
+use ssz_derive::{Decode, Encode};
+use ssz_types::VariableList;
+// TODO: Replace this with the actual max # of bytes an EE can have
+// Currently this is arbitrarily set to 256KB max size
```
Comment (villanuevawill) [inline eth2/types/src/execution_environment.rs:5]: makes sense

File: eth2/types/src/newtypes.rs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+use fixed_hash::construct_fixed_hash;
+
+// Necessary for impl_common macro
+use serde::{Deserialize, Serialize};
+use slog;
+use ssz::{ssz_encode, Decode, DecodeError, Encode};
+use std::cmp::{Ord, Ordering};
+use std::fmt;
+use std::hash::{Hash, Hasher};
+use std::ops::{Add, AddAssign, Div, DivAssign, Mul, MulAssign, Rem, Sub, SubAssign};
```
Comment (villanuevawill) [inline eth2/types/src/newtypes.rs]: Where are these used?

File: eth2/types/src/newtypes.rs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+// https://github.com/sigp/lighthouse repo with some small modifications.  Specifically, added new
+// types, and removed code not needed for this repo.
+// The analog of the `Root` struct in this repo is H256 in lighthouse (and in some other
+// ethereum-focused crates). Decided to use `Root` in this repo instead of `H256` because the spec
+// refers to the data type as a `Root`.  This can be changed in the future if necessary.
+use fixed_hash::construct_fixed_hash;
+
+// Necessary for impl_common macro
+use serde::{Deserialize, Serialize};
+use slog;
```
Comment (villanuevawill) [inline eth2/types/src/newtypes.rs]: We will install our own logging system, right? Where is this used?

File: eth2/types/src/newtypes.rs [hunk 7]
```diff
@@ -0,0 +1,88 @@
+// Many patterns copied from the slot_epoch.rs file in the
```
Comment (villanuevawill) [inline eth2/types/src/newtypes.rs]: Why is this called newtypes.rs?  You'll see I did a PR - https://github.com/quilt/lighthouse/pull/1/files#diff-97bf59d81db403d9da467c2c12a3e8cf which introduces slot_epoch patterns that are shard specific. You're going to need epoch conversions, etc. so I'd follow those patterns. Still unsure of why this would be called newTypes vs. just slot_epoch as well?

File: eth2/types/src/u64_macros.rs [hunk 8]
```diff
@@ -0,0 +1,610 @@
+// Copied from the slot_epoch_macros.rs file in https://github.com/sigp/lighthouse with some small
+// modifications.  Specifically, updated ssz macro to match the eth2_ssz crate's Encode trait, which
+// is slightly different than the Encode trait used internally in the sigp/lighthouse repo.
+
```
Comment (villanuevawill) [inline eth2/types/src/u64_macros.rs:4]: Did you make any changes to this? You should ask lighthouse to publish this in particular and just import it? Any rationale behind not doing that for pieces you do not alter?

File: eth2/types/src/eth_spec.rs [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+///
+/// Spec v0.9.1
+#[derive(Clone, PartialEq, Debug, Default, Serialize, Deserialize)]
+pub struct MainnetEthSpec;
+
+impl EthSpec for MainnetEthSpec {
+    type ShardCount = U64;
+    type MaxExecutionEnvironments = U65536;
+    type JustificationBitsLength = U4;
+    type MaxValidatorsPerCommittee = U2048;
```
Comment (villanuevawill) [inline eth2/types/src/eth_spec.rs]: Again, I'd pop out the ones we don't need for now and then you can add them slowly in. Also do we want a mainnetEthspec setup or the minimalistic one?

File: eth2/utils/serde_hex/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,80 @@
+// Copied from the file of the same name in the https://github.com/sigp/lighthouse repo with some
+// modifications and deletions.  Specifically, removed all code not needed for this repo.
+
+use hex;
```
Comment (villanuevawill) [inline eth2/utils/serde_hex/src/lib.rs:4]: What modifications did you make exactly? Can you please lay them out/point them out?

Comment (villanuevawill) [reply eth2/utils/serde_hex/src/lib.rs:4]: I'm not sure why you would need to make modifications and copy this over. That's my take.

**Review by villanuevawill — CHANGES_REQUESTED** [review]
A lot of these don't need modificaitons and we should ask lighthouse to publish these crates. In the meantime, I'd say it's fine to do this on a non-master branch. Lets coordinate this and have a chat with the lighthouse guys.

File: eth2/types/src/eth_spec.rs (hunk 3 above)
Comment (gjtrowbridge) [reply eth2/types/src/eth_spec.rs:2]: I think it's nice to have these--it has the whole spec in there, which is nice for seeing both full context on what values are defined in the spec and because if/when we need them we can use them.  One option that maybe is more attractive is I can comment out the parts that aren't actively in use yet (similar to what I did in BeaconState so folks can see the spec but don't need to worry about the commented out parts).  Open to thoughts here.

File: eth2/types/src/newtypes.rs (hunk 5 above)
Comment (gjtrowbridge) [reply eth2/types/src/newtypes.rs]: Necessary for the macros

File: eth2/types/src/newtypes.rs (hunk 6 above)
Comment (gjtrowbridge) [reply eth2/types/src/newtypes.rs]: Let me look, possible I can remove this.

File: eth2/types/src/newtypes.rs (hunk 7 above)
Comment (gjtrowbridge) [reply eth2/types/src/newtypes.rs]: called it newtypes because I have more than slot and epoch in there, and I didn't see a compelling reason to give those their own file.  I'll take a look at your PR, but so far no additional stuff is needed beyond what I have there, but may need more later for sure.

File: eth2/types/src/u64_macros.rs (hunk 8 above)
Comment (gjtrowbridge) [reply eth2/types/src/u64_macros.rs:4]: Yes, I did.  Details are in the comment.  I'm open to using a crate here but not sure how generally useful it will be so not sure it makes sense to push as a crate.  Either way works for me though

File: eth2/types/src/eth_spec.rs (hunk 9 above)
Comment (gjtrowbridge) [reply eth2/types/src/eth_spec.rs]: If you're cool with it I'll just comment 'em.  I think either ethspec works equally well at this point, it's very easy to change later if necessary.

File: eth2/utils/serde_hex/src/lib.rs (hunk 10 above)
Comment (gjtrowbridge) [reply eth2/utils/serde_hex/src/lib.rs:4]: it's used somewhere, let me recall where.  It's possible I can remove this and the thing that depends on it, though, I'll look.

File: eth2/types/src/eth_spec.rs (hunk 3 above)
Comment (villanuevawill) [reply eth2/types/src/eth_spec.rs:2]: Yeah, commenting out seems good from my end. It could be helpful for someone to dive in and see easily what is actually relevant in the current state.

File: eth2/types/src/newtypes.rs (hunk 5 above)
Comment (villanuevawill) [reply eth2/types/src/newtypes.rs]: ah that's right

File: eth2/types/src/newtypes.rs (hunk 6 above)
Comment (villanuevawill) [reply eth2/types/src/newtypes.rs]: Yeah - you don't need this. We'll insert our own logging - which I encourage you to do. But you mayh want to use slog. Up to you!

File: eth2/types/src/newtypes.rs (hunk 7 above)
Comment (villanuevawill) [reply eth2/types/src/newtypes.rs]: Yeah - I'd just avoid a generic name like that and just define it as slot_epoch.rs to stay in pattern.

File: eth2/types/src/eth_spec.rs (hunk 9 above)
Comment (villanuevawill) [reply eth2/types/src/eth_spec.rs]: yeah sounds good to me to just comment

File: eth2/utils/serde_hex/src/lib.rs (hunk 10 above)
Comment (gjtrowbridge) [reply eth2/utils/serde_hex/src/lib.rs:4]: Yeah it's necessary for some logic in the `ssz_types` package.  I do think we'll want to replace that package sooner rather than later with a crate, so for now I'm going to keep this in and we can replace it later.

File: eth2/types/src/newtypes.rs (hunk 7 above)
Comment (gjtrowbridge) [reply eth2/types/src/newtypes.rs]: Ok, I don't think I'm totally sold but I don't feel too strongly. Changed to `slot_epoch_root`, happy to change more in the future as necessary.

File: eth2/types/src/newtypes.rs (hunk 6 above)
Comment (gjtrowbridge) [reply eth2/types/src/newtypes.rs]: Good call, wasn't needed, removed it.

**Comment by gjtrowbridge** [comment]
Thanks for the reviews and comments, guys.  I believe I've addressed all outstanding comments, so for now I'll go ahead and land this.  If more changes become necessary in future, I'll update.  As requested, I'll land this to a dev branch for now so it's not on master until we have more of the copied logic being pulled in from crates vs. partially copied in.
